### ISSUE TITLE
u64::satuating_div is useless

### ISSUE THREAD

**Issue by ijackson** [rust-lang/rust#122821]
`pub const fn saturating_div(self, rhs: u64) -> u64`

> This function will panic if rhs is 0.

This makes no sense.  The *only* way that `u64 / u64` could overfllow is if `rhs = 0`.

The same is true for the other unsigned integer types.  The signed types have the excuse that they wouldn't know whether to count division by zero as saturating to -ve or +ve maximum.

**Comment by Nilstrieb** [comment]
I think these functions mostly exist for consistency

**Comment by ijackson** [comment]
That's all very well, but surely this one should return `u64::MAX` instead of panicking, on division by zero.

**Comment by Jules-Bertholet** [comment]
"Consistency" isn't usually sufficient justification for having a method. `u64::abs` doesn't exist, for example.

Perhaps we should just deprecate these?

**Comment by ryanavella** [comment]
> The signed types have the excuse that they wouldn't know whether to count division by zero as saturating to -ve or +ve maximum.

I don't think this was the only reason. On the PR (#87921) for the `std::num::Saturating` type, the original implementation guaranteed `x / 0 == MAX`, but that was changed when @kennytm noted it could give surprising results for `0 / 0`.

It's also worth noting that if `u64::saturating_div` saturated to `u64::MAX` instead of panicking, that it is somewhat redundant with `u64::checked_div` (the more general of the two)

**Comment by kennytm** [comment]
>  The signed types have the excuse that they wouldn't know whether to count division by zero as saturating to -ve or +ve maximum.

No, for signed type `saturating_div` makes sense because `i32::MIN / -1` overflows. It has nothing to do with divide-by-zero (which also panics).

If `u64::saturating_div` is useless because `x / 0` panics then so is `u64::wrapping_div` and `u64::overflowing_div` and `u64::strict_div` which are all equivalent.

**Comment by Jules-Bertholet** [comment]
> If `u64::saturating_div` is useless because `x / 0` panics

It's not useless just because of this, it's useless because it has identical behavior to `u64::div` in all cases. Noone is proposing deprecating or changing `saturating_div` for the signed integers. (See my previous analogy to `abs`)

**Comment by kennytm** [comment]
@Jules-Bertholet All of `u64::wrapping_div` and `u64::strict_div` are identical to `u64::div` and `u64::overflowing_div` is just `(x/y, false)`.

How is a non-existing function like `u64::abs` relevant? The consistency of `u64::saturating_div` is with `u64::saturating_{add,sub,mul,...}`, *not* `i64::saturating_div`.

**Comment by Jules-Bertholet** [comment]
> All of `u64::wrapping_div` and `u64::strict_div` are identical to `u64::div`, and `u64::overflowing_div` is just `(x/y, false)`.

Then IMO these should be deprecated also.

**Comment by AaronKutch** [comment]
No it is not useless, consistency in integer methods is extremely useful for macro writers. The only issue is that `saturating_div` possibly should have used the RISC-V convention where division by zero results in `iX::MIN` if the dividend is negative, 0 if zero, and `iX::MAX` if positive.

**Comment by Noratrieb** [comment]
Yeah, having this consistent integer methods makes writing macros easier (and it makes writing the standard library implementation macros easier :3), and they don't hurt. I think this should be closed.

**Comment by Pr0methean** [comment]
> No it is not useless, consistency in integer methods is extremely useful for macro writers. The only issue is that `saturating_div` possibly should have used the RISC-V convention where division by zero results in `iX::MIN` if the dividend is negative, 0 if zero, and `iX::MAX` if positive.

That's also what you'd get if you did the division in floating point and then converted the result (of infinity or NaN) back.
